File: sample-projects/go/adapter/persistence/persisted_creatures.go
package persistence

import (
	"de.sots/cellarsandcentaurs/application"
	"de.sots/cellarsandcentaurs/domain/model"
	"de.sots/cellarsandcentaurs/domain/service"
)

var _ service.Creatures = (*PersistedCreatures)(nil)

type PersistedCreatures struct {
	repository *CreatureRepository
}

func NewPersistedCreatures(repository *CreatureRepository) *PersistedCreatures {
	return &PersistedCreatures{repository: repository}
}

func (p *PersistedCreatures) Save(creature *model.Creature) error {
	p.repository.SaveCreature(creature)
	return nil
}

func (p *PersistedCreatures) Find(id model.CreatureId) (*model.Creature, error) {
	entity, found := p.repository.FindOne(id.String())
	if !found {
		return nil, model.NewNoSuchCreatureException(id)
	}
	return model.NewCreatureOfType(model.CreatureIdOf(entity.Id), application.StandardCreatureType), nil
}

File: sample-projects/c/src/adapter/persistence/persisted_creatures.c
#include <stdio.h>

#include "../../../include/de/sots/cellarsandcentaurs/adapter/persistence/persisted_creatures.h"
#include "../../../include/de/sots/cellarsandcentaurs/adapter/persistence/creature_entity.h"
#include "../../../include/de/sots/cellarsandcentaurs/domain/model/creature.h"
#include "../../../include/de/sots/cellarsandcentaurs/domain/model/creature_id.h"
#include "../../../include/de/sots/cellarsandcentaurs/domain/model/no_such_creature_exception.h"
#include "../../../include/de/sots/cellarsandcentaurs/application/application.h"

static void persisted_creatures_save(void *self, Creature *creature)
{
    PersistedCreatures *persisted = self;
    creature_repository_save(persisted->repository, creature_entity_new(creature_get_id(creature)->id));
}

static Creature *persisted_creatures_find(void *self, CreatureId id)
{
    PersistedCreatures *persisted = self;
    CreatureEntity *entity = creature_repository_find_one(persisted->repository, id.id);
    if (entity == NULL) {
        NoSuchCreatureException exception = no_such_creature_exception_new(id);
        fputs(no_such_creature_exception_message(&exception), stderr);
        return NULL;
    }
    return creature_new(creature_id_new(entity->id), CREATURE_FACADE_STANDARD_CREATURE_TYPE);
}

PersistedCreatures persisted_creatures_new(CreatureRepository *repository)
{
    PersistedCreatures persisted;
    persisted.repository = repository;
    persisted.port.self = &persisted;
    persisted.port.save = persisted_creatures_save;
    persisted.port.find = persisted_creatures_find;
    return persisted;
}

Creatures *persisted_creatures_as_creatures(PersistedCreatures *persisted)
{
    return &persisted->port;
}
